pkg/kubelet/pleg/generic.go

```diff
@@ -116,40 +158,191 @@ func (g *GenericPLEG) relist() {
+
+		// Process all currently visible containers in the pod.
+		for _, container := range pod.Containers {
+			cid := container.ID
+			oldState := getContainerState(oldPod, cid)
+			newState := convertState(container.State)
+			e := generateEvent(pod.ID, cid.ID, oldState, newState)
+			updateEvents(eventsByPodID, e)
+		}
+
+		if oldPod == nil {
+			continue
+		}
+		// Process all containers in the old pod, but no longer in the new pod.
+		for _, oldContainer := range oldPod.Containers {
+			cid := oldContainer.ID
+			oldState := convertState(oldContainer.State)
+			newState := getContainerState(pod, cid)
+			if newState != plegContainerNonExistent {
+				// We already processed the container.
+				continue
+			}
+			// Container no longer visible, generate an event.
+			e := generateEvent(pod.ID, cid.ID, oldState, plegContainerNonExistent)
+			updateEvents(eventsByPodID, e)
+		}
+	}
+	// Process all pods that are no longer visible.
+	for pid := range g.podRecords {
+		if pod := g.podRecords.getCurrent(pid); pod != nil {
+			continue
+		}
+		oldPod := g.podRecords.getOld(pid)
+		for _, oldContainer := range oldPod.Containers {
+			cid := oldContainer.ID
+			oldState := convertState(oldContainer.State)
+			e := generateEvent(oldPod.ID, cid.ID, oldState, plegContainerNonExistent)
+			updateEvents(eventsByPodID, e)
+		}
+	}
+
+	if g.cacheEnabled() {
+		// If there are events associated with a pod, we should update the
+		// podCache.
+		for pid, events := range eventsByPodID {
+			pod := g.podRecords.getCurrent(pid)
+			if pod == nil {
+				// The pod is missing in the current relist. This means that
+				// the pod has no visible (active or inactive) containers.
+				g.removeFromCache(pid)
+				continue
+			}
+			// updateCache() will inspect the pod and update the cache. If an
+			// error occurs during the inspection, we want PLEG to retry again
+			// in the next relist. To achieve this, we do not update the
+			// associated podRecord of the pod, so that the change will be
+			// detect again in the next reilst.
+			// TODO: If many pods changed during the same relist period,
+			// inspecting the pod and getting the PodStatus to update the cache
+			// serially may take a while. We should be aware of this and
+			// parallelize if needed.
+			if err := g.updateCache(pod); err != nil {
+				glog.Errorf("PLEG: Ignoring events for pod %s/%s: %v", pod.Name, pod.Namespace, err)
+			} else {
+				// Update the internal storage and send out the events.
+				g.podRecords.update(pid)
+				for i := range events {
+					g.eventChannel <- events[i]
+				}
 			}
-			// Generate an event if required.
-			glog.V(7).Infof("GenericPLEG: %v/%v: %v -> %v", p.ID, cid, oldState, c.State)
-			if e := generateEvent(p.ID, cid, oldState, c.State); e != nil {
-				events = append(events, e)
+		}
+		// Update the cache timestamp.  This needs to happen *after*
+		// all pods have been properly updated in the cache.
+		g.cache.SetGlobalTimestamp(timestamp)
+	} else {
```

Comment:
I think it's more clear to keep the if-else blocks since `return` the blocks are quite big -- people may miss the return statement.